SerialPrograms/Source/PokemonSwSh/Programs/DateSpamFarmers/PokemonSwSh_DateSpam-BerryFarmer2.cpp

```diff
@@ -0,0 +1,275 @@
+    PA_ADD_OPTION(FETCH_ATTEMPTS);
+    PA_ADD_OPTION(SAVE_ITERATIONS);
+
+    PA_ADD_STATIC(m_advanced_options);
+    PA_ADD_OPTION(EXIT_BATTLE_TIMEOUT);
+    PA_ADD_OPTION(START_BATTLE_DELAY);
+    PA_ADD_OPTION(RUSTLING_INTERVAL);
+    PA_ADD_OPTION(RUSTLING_TIMEOUT);
+    PA_ADD_OPTION(SECONDARY_ATTEMPT_MASH_TIME);
+}
+
+
+BerryFarmer2::Rustling BerryFarmer2::check_rustling(SingleSwitchProgramEnvironment& env, BotBaseContext& context) {
+    // wait some time in order to not detect rustling from previous fetch attempt
+    pbf_wait(context, 80);
+    context.wait_for_all_requests();
+
+    std::chrono::time_point initial_rustling_time = std::chrono::system_clock::now();
+
+    BerryTreeRustlingSoundDetector initial_rustling_detector(env.console.logger(), env.console, [&](float error_coefficient) -> bool {
+        //  Warning: This callback will be run from a different thread than this function.
+        return true;
+        });
+
+    BerryTreeRustlingSoundDetector secondary_rustling_detector(env.console.logger(), env.console, [&](float error_coefficient) -> bool {
+        //  Warning: This callback will be run from a different thread than this function.
+        return true;
+        });
+
+    StandardBattleMenuWatcher battle_menu_detector(false);
+    StartBattleWatcher start_battle_detector;
+
+    Rustling result = Rustling::No;
+    int ret = run_until(
+        env.console, context,
+        [&](BotBaseContext& context) {
+            pbf_wait(context, RUSTLING_TIMEOUT);
+            context.wait_for_all_requests();
+        },
+        { {initial_rustling_detector}, {battle_menu_detector}, {start_battle_detector} }
+        );
+
+    if (ret == 0) {
+        env.console.log("BerryFarmer: Initial Rustling detected.");
+        initial_rustling_time = std::chrono::system_clock::now();
+
+        result = Rustling::Slow;
+        
+        int ret1 = run_until(
+            env.console, context,
+            [&](BotBaseContext& context) {
+                pbf_wait(context, RUSTLING_TIMEOUT);
+                context.wait_for_all_requests();
+            },
+            { {secondary_rustling_detector} }
+            );
+
+        if (ret1 == 0) {
+            env.console.log("BerryFarmer: Secondary Rustling detected.");
+            std::chrono::time_point secondary_rustling_time = std::chrono::system_clock::now();
+            if (std::chrono::duration_cast<std::chrono::milliseconds>(secondary_rustling_time - initial_rustling_time).count() <= RUSTLING_INTERVAL) {
+                result = Rustling::Fast;
+            }
+        }
+
+    }
+    else if (ret == 1) {
+        env.console.log("BerryFarmer: Battle Menu detected!");
+        result = Rustling::Battle;
+    } 
+    else if (ret == 2) {
+        env.console.log("BerryFarmer: Battle Start detected.");
+        // waiting here so we can be sure the battle is fully underway so escaping is possible
+        pbf_wait(context, START_BATTLE_DELAY);
```

Comment:
Is this case rare? Blind waiting for the battle to fully start seems suboptimal? There's a `StandardBattleMenuWatcher` that will wait for it with feedback.

Reply:
A Battle should be quite rare. Changed to `StandardBattleMenuWatcher` anyway since it was already used elswehere.